pkg/kubefed2/unjoin.go

```diff
@@ -0,0 +1,409 @@
+		return nil
+	}
+
+	err := crClientset.ClusterregistryV1alpha1().Clusters(controllerutil.MulticlusterPublicNamespace).Delete(unjoiningClusterName,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// deleteFederatedClusterAndSecret deletes a federated cluster resource that associates
+// the cluster and secret.
+func deleteFederatedClusterAndSecret(hostClientset client.Interface, fedClientset *fedclient.Clientset,
+	federationNamespace, unjoiningClusterName string, dryRun bool) (*fedv1a1.FederatedCluster, error) {
+	if dryRun {
+		return nil, nil
+	}
+
+	fedCluster, err := fedClientset.CoreV1alpha1().FederatedClusters(federationNamespace).Get(
+		unjoiningClusterName, metav1.GetOptions{})
+	if err != nil {
+		return fedCluster, err
+	}
+
+	err = fedClientset.CoreV1alpha1().FederatedClusters(federationNamespace).Delete(
+		unjoiningClusterName, &metav1.DeleteOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = hostClientset.CoreV1().Secrets(federationNamespace).Delete(fedCluster.Spec.SecretRef.Name,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// deleteRBACResources deletes the cluster role, cluster rolebindings and service account
+// from the unjoining cluster.
+func deleteRBACResources(hostClusterClientset, unjoiningClusterClientset client.Interface,
+	namespace, unjoiningClusterName, hostClusterName string, dryRun bool) error {
+
+	saName := util.ClusterServiceAccountName(unjoiningClusterName, hostClusterName)
+
+	glog.V(2).Infof("Deleting cluster role binding for service account: %s in unjoining cluster: %s",
+		saName, unjoiningClusterName)
+
+	err := deleteClusterRoleAndBinding(unjoiningClusterClientset, saName, namespace,
+		unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Error deleting cluster role binding for service account: %s in unjoining cluster: %v",
+			saName, err)
+		return err
+	}
+
+	glog.V(2).Infof("Deleted cluster role binding for service account: %s in unjoining cluster: %s",
+		saName, unjoiningClusterName)
+
+	glog.V(2).Infof("Deleting service account %s in unjoining cluster: %s", saName, unjoiningClusterName)
+
+	err = deleteServiceAccount(unjoiningClusterClientset, saName, namespace, dryRun)
+	if err != nil {
+		return err
+	}
+
+	glog.V(2).Infof("Deleted service account %s in unjoining cluster: %s", saName, unjoiningClusterName)
+
+	return nil
+}
+
+// deleteFedNSFromUnjoinCluster deletes the federation namespace and multi cluster
+// public namespace in the unjoining cluster.
+func deleteFedNSFromUnjoinCluster(unjoiningClusterClientset client.Interface,
+	federationNamespace, unjoiningClusterName, hostClusterName string, dryRun bool) error {
+
+	if dryRun {
+		return nil
+	}
+
+	err := unjoiningClusterClientset.CoreV1().Namespaces().Delete(federationNamespace,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		glog.V(2).Infof("Could not delete federation namespace: %s due to: %v", federationNamespace, err)
```

Comment:
(No action required) The more common logging pattern in fedv2 is `Error performing action on resource %q: %s`, avoiding the need for 'due to' and multiple colons.